Cannot set port mapping via property

Given the following pom, I'd expect the flyway plugin to be able to run the migrate task. 

This pom is a simplified version of a pom that works but the working version cannot set postgres_user and postgres_password environment variables. 

So I wanted to use the external property support to set these variables but now the port mapping does not work. When I do 

`mvn verify `

based on the given pom file:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <groupId>mytest</groupId>
    <artifactId>development</artifactId>
    <version>1.0-SNAPSHOT</version>

    <properties>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
        <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>

        <postgres.docker.name>postgres:latest</postgres.docker.name>
        <postgres.docker.env.POSTGRES_USER>postgres</postgres.docker.env.POSTGRES_USER>
        <postgres.docker.env.POSTGRES_PASSWORD>postgres</postgres.docker.env.POSTGRES_PASSWORD>
        <postgres.docker.wait.time>11000</postgres.docker.wait.time>
        <postgres.docker.ports.1>18800:5432</postgres.docker.ports.1>
    </properties>

    <build>
        <plugins>

            <!--DOCKER-->
            <plugin>
                <groupId>io.fabric8</groupId>
                <artifactId>docker-maven-plugin</artifactId>
                <version>0.18.1</version>

                <configuration>
                    <images>
                        <image>
                            <alias>simple</alias>
                            <external>
                                <type>properties</type>
                                <prefix>postgres.docker</prefix>
                            </external>
                            <run>
                                <namingStrategy>alias</namingStrategy>
                            </run>
                        </image>
                    </images>
                </configuration>

                <executions>
                    <execution>
                        <id>start</id>
                        <phase>pre-integration-test</phase>
                        <goals>
                            <goal>build</goal>
                            <goal>start</goal>
                        </goals>
                    </execution>
                    <execution>
                        <id>stop</id>
                        <phase>post-integration-test</phase>
                        <goals>
                            <goal>stop</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>

            <!--FLYWAY -->
            <plugin>
                <groupId>org.flywaydb</groupId>
                <artifactId>flyway-maven-plugin</artifactId>
                <version>4.1.1</version>
                <configuration>
                    <url>jdbc:postgresql://127.0.0.1:18800/postgres</url>
                    <user>postgres</user>
                    <password>postgres</password>
                </configuration>
                <executions>
                    <execution>
                        <id>start</id>
                        <phase>pre-integration-test</phase>
                        <goals>
                            <goal>clean</goal>
                            <goal>migrate</goal>
                        </goals>
                    </execution>
                </executions>
                <dependencies>
                    <dependency>
                        <groupId>org.postgresql</groupId>
                        <artifactId>postgresql</artifactId>
                        <version>42.0.0.jre7</version>
                    </dependency>
                </dependencies>
            </plugin>
        </plugins>
    </build>

    <dependencies>
        <!--JUNIT-->
        <dependency>
            <groupId>junit</groupId>
            <artifactId>junit</artifactId>
            <version>4.12</version>
        </dependency>

        <!--MOCKITO-->
        <dependency>
            <groupId>org.mockito</groupId>
            <artifactId>mockito-core</artifactId>
            <version>2.7.16</version>
        </dependency>

        <!--HIBERNATE-->
        <dependency>
           <groupId>org.hibernate</groupId>
           <artifactId>hibernate-core</artifactId>
           <version>5.2.8.Final</version>
        </dependency>

        <!--POSTGRES JDBC-->
        <dependency>
            <groupId>org.postgresql</groupId>
            <artifactId>postgresql</artifactId>
            <version>42.0.0.jre7</version>
        </dependency>
    </dependencies>
</project>
```

Flyway cannot connect to the container, producing the following exception:

```
[ERROR] Failed to execute goal org.flywaydb:flyway-maven-plugin:4.1.1:clean (start) on project development: org.flywaydb.core.internal.dbsupport.FlywaySqlException:
[ERROR] Unable to obtain Jdbc connection from DataSource (jdbc:postgresql://127.0.0.1:18800/postgres) for user 'postgres': Connection to 127.0.0.1:18800 refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.
[ERROR] ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
[ERROR] SQL State  : 08001
[ERROR] Error Code : 0
[ERROR] Message    : Connection to 127.0.0.1:18800 refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections. Connection refused (Connection refused)
```

I can see that the waiting time property works, but I could not get port mapping to work. If I don't use external properties, then I can't set postgres user/pass and it ends up allowing any password access with postgres user which I don't like.

What am I missing here?

--- a/doc/changelog.md
+++ b/doc/changelog.md
@@ -5,6 +5,7 @@
   - Allow image with multiple path segments (#694)
   - Add support for PKCS#8 private keys in pem.key file. (#730)
   - Improve resource management for certificates and keys. (#730)
+  - When using properties for configuration only build when `from` or `fromExt` is set (#736)
 
 * **0.20.0** (2017-02-17)
   - Removed `build-nofork` and `source-nofork` in favor for a more direct solution which prevents forking of the lifecycle. Please refer the documentation, chapter "Assembly" for more information about this.

--- a/src/main/asciidoc/inc/external/_property_configuration.adoc
+++ b/src/main/asciidoc/inc/external/_property_configuration.adoc
@@ -2,7 +2,7 @@
 For simple needs the image configuration can be completely defined via
 Maven properties which are defined outside of this plugin's
 configuration. Such a property based configuration can be selected
-with an `<type>` of `props`. As extra configuration a prefix for the
+with an `<type>` of `properties`. As extra configuration a prefix for the
 properties can be defined which by default is `docker`.
 
 .Example
@@ -10,7 +10,7 @@ properties can be defined which by default is `docker`.
 ----
 <image>
   <external>
-     <type>props</type>
+     <type>properties</type>
      <prefix>docker</prefix> <!-- this is the default -->
   </external>
 </image>
@@ -18,7 +18,8 @@ properties can be defined which by default is `docker`.
 
 Given this example configuration a single image configuration is built
 up from the following properties, which correspond to the corresponding
-values in the `<build>` and `<run>` sections.
+values in the `<build>` and `<run>` sections. A build configuration is only created
+when a `docker.from` or a `docker.fromExt` is set.
 
 .External properties
 [cols="1,5"]
@@ -92,11 +93,8 @@ values in the `<build>` and `<run>` sections.
 | *docker.exposedPropertyKey*
 | Property part for the exposed container properties like internal IP addresses as described in <<start-overview, docker:start>>.
 
-| *docker.workdir*
-| Container working directory
-
 | *docker.env.VARIABLE*
-| Sets an environment variable. E.g. `<docker.env.JAVA_OPTS>-Xmx512m</docker.env.JAVA_OPTS>` sets the environment variable `JAVA_OPTS`. Multiple such entries can be provided. This environment is used both for building images and running containers. The value cannot be empty but can contain Maven property names which are resolved before the Dockerfile is created
+| Sets an environment variable. E.g. `<docker.env.JAVA_OPTS>-Xmx512m</docker.env.JAVA_OPTS>` sets the environment variable `JAVA_OPTS`. Multiple such entries can be provided. This environment is used both for building images and running containers. The value cannot be empty but can contain Maven property names which are resolved before the Dockerfile is created.
 
 | *docker.envPropertyFile*
 | specifies the path to a property file whose properties are used as environment variables. The environment variables takes precedence over any other environment variables specified.
@@ -105,10 +103,10 @@ values in the `<build>` and `<run>` sections.
 | List of `host:ip` to add to `/etc/hosts`
 
 | *docker.from*
-| Base image for building an image
+| Base image for building an image. Must be set when an image is created (or `fromExt`)
 
 | *docker.fromExt.VARIABLE*
-| Base image for building an image (extended format)
+| Base image for building an image (extended format), which also triggers a build of an image.
 
 | *docker.healthcheck.cmd*
 | Command to use for a healthcheck
@@ -254,8 +252,11 @@ values in the `<build>` and `<run>` sections.
 | *docker.wait.kill*
 | Time in milliseconds to wait between sending SIGTERM and SIGKILL to a container when stopping it.
 
+| *docker.workdir*
+| Container working directory where the image is build in
+
 | *docker.workingDir*
-| Working dir for commands to run in
+| Current Working dir for commands to run in when running containers
 |===
 
 Any other `<run>` or `<build>` sections are ignored when this handler
@@ -288,7 +289,7 @@ run and build configuration sections.
         <images>
           <image>
             <external>
-              <type>props</type>
+              <type>properties</type>
               <prefix>docker</prefix>
             </external>
           </image>

--- a/src/main/java/io/fabric8/maven/docker/AbstractDockerMojo.java
+++ b/src/main/java/io/fabric8/maven/docker/AbstractDockerMojo.java
@@ -222,16 +222,24 @@ public abstract class AbstractDockerMojo extends AbstractMojo implements Context
                 ServiceHub serviceHub = serviceHubFactory.createServiceHub(project, session, access, log, logSpecFactory);
                 executeInternal(serviceHub);
             } catch (DockerAccessException exp) {
-                log.error("%s", exp.getMessage());
+                logException(exp);
                 throw new MojoExecutionException(log.errorMessage(exp.getMessage()), exp);
             } catch (MojoExecutionException exp) {
-                log.error("%s", exp.getMessage());
+                logException(exp);
                 throw exp;
             } finally {
                 if (access != null) {
                     access.shutdown();
                 }
             }
+        }
+    }
+
+    private void logException(Exception exp) {
+        if (exp.getCause() != null) {
+            log.error("%s [%s]", exp.getMessage(), exp.getCause().getMessage());
+        } else {
+            log.error("%s", exp.getMessage());
         }
     }
 

--- a/src/main/java/io/fabric8/maven/docker/StartMojo.java
+++ b/src/main/java/io/fabric8/maven/docker/StartMojo.java
@@ -195,7 +195,7 @@ public class StartMojo extends AbstractDockerMojo {
             Thread.currentThread().interrupt();
             throw new MojoExecutionException("interrupted", e);
         } catch (IOException e) {
-            throw new MojoExecutionException("I/O Error",e);
+            throw new MojoExecutionException("I/O Error", e);
         } finally {
             shutdownExecutorService(executorService);
 

--- a/src/main/java/io/fabric8/maven/docker/access/DockerAccessException.java
+++ b/src/main/java/io/fabric8/maven/docker/access/DockerAccessException.java
@@ -23,12 +23,21 @@ public class DockerAccessException extends IOException {
     public DockerAccessException(String message) {
         super(message);
     }
-    
+
     public DockerAccessException(String format, Object...args) {
         super(String.format(format, args));
     }
 
     public DockerAccessException(Throwable cause, String format, Object ... args) {
         super(String.format(format, args),cause);
+    }
+
+    @Override
+    public String getMessage() {
+        if (getCause() != null) {
+            return String.format("%s : %s", super.getMessage(), getCause().getMessage());
+        } else {
+            return super.getMessage();
+        }
     }
 }

--- a/src/main/java/io/fabric8/maven/docker/config/handler/property/PropertyConfigHandler.java
+++ b/src/main/java/io/fabric8/maven/docker/config/handler/property/PropertyConfigHandler.java
@@ -64,7 +64,17 @@ public class PropertyConfigHandler implements ExternalConfigHandler {
                         .build());
     }
 
+    // Enable build config only when a `.from.` is configured
+    private boolean buildConfigured(String prefix, Properties properties) {
+        return withPrefix(prefix, FROM, properties) != null ||
+               mapWithPrefix(prefix,FROM_EXT,properties) != null;
+    }
+
+
     private BuildImageConfiguration extractBuildConfiguration(String prefix, Properties properties) {
+        if (!buildConfigured(prefix, properties)) {
+            return null;
+        }
         return new BuildImageConfiguration.Builder()
                 .cmd(withPrefix(prefix, CMD, properties))
                 .cleanup(withPrefix(prefix, CLEANUP, properties))

--- a/src/test/java/io/fabric8/maven/docker/config/handler/property/PropertyConfigHandlerTest.java
+++ b/src/test/java/io/fabric8/maven/docker/config/handler/property/PropertyConfigHandlerTest.java
@@ -54,7 +54,7 @@ public class PropertyConfigHandlerTest extends AbstractConfigHandlerTest {
         assertFalse(resolveExternalImageConfig(getSkipTestData(ConfigKey.SKIP_BUILD, false)).getBuildConfiguration().skip());
         assertTrue(resolveExternalImageConfig(getSkipTestData(ConfigKey.SKIP_BUILD, true)).getBuildConfiguration().skip());
 
-        assertFalse(resolveExternalImageConfig(new String[] {k(ConfigKey.NAME), "image"}).getBuildConfiguration().skip());
+        assertFalse(resolveExternalImageConfig(new String[] {k(ConfigKey.NAME), "image", k(ConfigKey.FROM), "busybox"}).getBuildConfiguration().skip());
     }
 
     @Test
@@ -82,7 +82,8 @@ public class PropertyConfigHandlerTest extends AbstractConfigHandlerTest {
                         "docker.name","demo",
                         "docker.ports.1", "jolokia.port:8080",
                         "docker.ports.2", "9090",
-                        "docker.ports.3", "0.0.0.0:80:80"
+                        "docker.ports.3", "0.0.0.0:80:80",
+                        "docker.from", "busybox"
                                         ));
         assertEquals(1,configs.size());
         RunImageConfiguration runConfig = configs.get(0).getRunConfiguration();
@@ -102,6 +103,7 @@ public class PropertyConfigHandlerTest extends AbstractConfigHandlerTest {
     public void testRunCommands() {
         List<ImageConfiguration> configs = resolveImage(
             imageConfiguration,props(
+                "docker.from", "base",
                 "docker.name","demo",
                 "docker.run.1", "foo",
                 "docker.run.2", "bar",
@@ -119,6 +121,7 @@ public class PropertyConfigHandlerTest extends AbstractConfigHandlerTest {
     public void testEnvAndLabels() throws Exception {
         List<ImageConfiguration> configs = resolveImage(
                 imageConfiguration,props(
+                    "docker.from", "baase",
                         "docker.name","demo",
                         "docker.env.HOME", "/tmp",
                         "docker.env.root.dir", "/bla",
@@ -164,15 +167,23 @@ public class PropertyConfigHandlerTest extends AbstractConfigHandlerTest {
 
     @Test
     public void testNoCleanup() throws Exception {
-        String[] testData = new String[] {k(ConfigKey.NAME), "image", k(ConfigKey.CLEANUP), "none" };
+        String[] testData = new String[] {k(ConfigKey.NAME), "image", k(ConfigKey.CLEANUP), "none", k(ConfigKey.FROM), "base" };
 
         ImageConfiguration config = resolveExternalImageConfig(testData);
         assertEquals(CleanupMode.NONE, config.getBuildConfiguration().cleanupMode());
     }
 
     @Test
+    public void testNoBuildConfig() throws Exception {
+        String[] testData = new String[] {k(ConfigKey.NAME), "image" };
+
+        ImageConfiguration config = resolveExternalImageConfig(testData);
+        assertNull(config.getBuildConfiguration());
+    }
+
+    @Test
     public void testDockerfile() throws Exception {
-        String[] testData = new String[] { k(ConfigKey.NAME), "image", k(ConfigKey.DOCKER_FILE_DIR), "src/main/docker/" };
+        String[] testData = new String[] { k(ConfigKey.NAME), "image", k(ConfigKey.DOCKER_FILE_DIR), "src/main/docker/", k(ConfigKey.FROM), "busybox" };
         ImageConfiguration config = resolveExternalImageConfig(testData);
         config.initAndValidate(ConfigHelper.NameFormatter.IDENTITY, null);
         assertTrue(config.getBuildConfiguration().isDockerFileMode());
@@ -181,7 +192,7 @@ public class PropertyConfigHandlerTest extends AbstractConfigHandlerTest {
 
     @Test
     public void testDockerArchive() {
-        String[] testData = new String[] { k(ConfigKey.NAME), "image", k(ConfigKey.DOCKER_ARCHIVE), "dockerLoad.tar" };
+        String[] testData = new String[] { k(ConfigKey.NAME), "image", k(ConfigKey.DOCKER_ARCHIVE), "dockerLoad.tar", k(ConfigKey.FROM), "busybox" };
         ImageConfiguration config = resolveExternalImageConfig(testData);
         config.initAndValidate(ConfigHelper.NameFormatter.IDENTITY, null);
         assertFalse(config.getBuildConfiguration().isDockerFileMode());
@@ -190,14 +201,14 @@ public class PropertyConfigHandlerTest extends AbstractConfigHandlerTest {
 
     @Test(expected = IllegalArgumentException.class)
     public void testInvalidDockerFileArchiveConfig() {
-        String[] testData = new String[] { k(ConfigKey.NAME), "image", k(ConfigKey.DOCKER_FILE_DIR),  "src/main/docker/", k(ConfigKey.DOCKER_ARCHIVE), "dockerLoad.tar" };
+        String[] testData = new String[] { k(ConfigKey.NAME), "image", k(ConfigKey.DOCKER_FILE_DIR),  "src/main/docker/", k(ConfigKey.DOCKER_ARCHIVE), "dockerLoad.tar", k(ConfigKey.FROM), "base" };
         ImageConfiguration config = resolveExternalImageConfig(testData);
         config.initAndValidate(ConfigHelper.NameFormatter.IDENTITY, null);
     }
 
     @Test
     public void testNoCacheDisabled() throws Exception {
-        String[] testData = new String[] {k(ConfigKey.NAME), "image", k(ConfigKey.NOCACHE), "false" };
+        String[] testData = new String[] {k(ConfigKey.NAME), "image", k(ConfigKey.NOCACHE), "false", k(ConfigKey.FROM), "base" };
 
         ImageConfiguration config = resolveExternalImageConfig(testData);
         assertEquals(false, config.getBuildConfiguration().nocache());
@@ -205,7 +216,7 @@ public class PropertyConfigHandlerTest extends AbstractConfigHandlerTest {
 
     @Test
     public void testNoCacheEnabled() throws Exception {
-        String[] testData = new String[] {k(ConfigKey.NAME), "image", k(ConfigKey.NOCACHE), "true" };
+        String[] testData = new String[] {k(ConfigKey.NAME), "image", k(ConfigKey.NOCACHE), "true", k(ConfigKey.FROM), "base" };
 
         ImageConfiguration config = resolveExternalImageConfig(testData);
         assertEquals(true, config.getBuildConfiguration().nocache());
@@ -213,7 +224,7 @@ public class PropertyConfigHandlerTest extends AbstractConfigHandlerTest {
 
     @Test
     public void testNoOptimise() throws Exception {
-        String[] testData = new String[] {k(ConfigKey.NAME), "image", k(ConfigKey.OPTIMISE), "false" };
+        String[] testData = new String[] {k(ConfigKey.NAME), "image", k(ConfigKey.OPTIMISE), "false", k(ConfigKey.FROM), "base" };
 
         ImageConfiguration config = resolveExternalImageConfig(testData);
         assertEquals(false, config.getBuildConfiguration().optimise());
@@ -275,7 +286,6 @@ public class PropertyConfigHandlerTest extends AbstractConfigHandlerTest {
         external.put("type", "props");
 
         ImageConfiguration config = new ImageConfiguration.Builder().name("image").alias("alias").externalConfig(external).build();
-        PropertyConfigHandler handler = new PropertyConfigHandler();
 
         List<ImageConfiguration> resolvedImageConfigs = resolveImage(config, props(testData));
         assertEquals(1, resolvedImageConfigs.size());
@@ -396,6 +406,7 @@ public class PropertyConfigHandlerTest extends AbstractConfigHandlerTest {
 
     private String[] getTestAssemblyData() {
         return new String[] {
+            k(ConfigKey.FROM), "busybox",
             k(ConfigKey.ASSEMBLY_BASEDIR), "/basedir",
             k(ConfigKey.ASSEMBLY_DESCRIPTOR_REF), "project",
             k(ConfigKey.ASSEMBLY_EXPORT_BASEDIR), "false",
@@ -466,7 +477,7 @@ public class PropertyConfigHandlerTest extends AbstractConfigHandlerTest {
     }
 
     private String[] getSkipTestData(ConfigKey key, boolean value) {
-        return new String[] {k(ConfigKey.NAME), "image", k(key), String.valueOf(value) };
+        return new String[] {k(ConfigKey.NAME), "image", k(key), String.valueOf(value), k(ConfigKey.FROM), "busybox" };
     }
 
     private String k(ConfigKey from) {

